src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt

```diff
@@ -175,4 +185,44 @@ class ProjectService(
 
             projectMembersWithUsers.toGrpcProjectMembers()
         }
+
+    override suspend fun getProjectInformation(request: GrpcProject.Information.Get): GrpcProject.Information =
+        withUser(userRepo) { currentUser ->
+            val projectId = parseUUID(request.projectId, EntityType.PROJECT)
+            val project = repo.getProjectById(projectId).getOrThrow()
+            ensureCurrentUserIsProjectMember(projectMemberRepo, projectId, currentUser)
+            val progress = projectPaperRepo.getProjectProgress(projectId)
+
+            val has = if (request.hasMask()) {
+                val fieldMaskPaths = FieldMaskUtil.normalize(request.mask).pathsList.toSet();
+                { path: String -> path in fieldMaskPaths }
+            } else {
+                { _ -> true }
+            }
+
+            GrpcProject.Information
+                .newBuilder()
+                .also {
+                    if (has("project_progress")) {
+                        it.setProjectProgress(progress)
+                    }
+                    if (has("creation_date")) {
+                        it.setCreationDate(
+                            Timestamp
+                                .newBuilder()
+                                .setSeconds(project.createdAt.toEpochSecond())
+                                .build(),
+                        )
+                    }
+                    if (has("last_stage_created")) {
+                        it.setLastStageStarted(
+                            Timestamp
+                                .newBuilder()
+                                .setSeconds(project.currentStageStartedAt.toEpochSecond())
+                                .build(),
+                        )
+                    }
+                }
+                .build()
```

Comment:
What's your opinion on the following? The timestamp builders can be replaced with the DSL, and we can decrease the indentation by just storing and using the builder. It's "last_stage_started" instead of "last_stage_created". Furthermore, we can just check whether a path is in the paths set:
```suggestion
            val paths = FieldMaskUtil.normalize(request.mask).pathsList.toSet();
            val builder = GrpcProject.Information.newBuilder()

            if ("project_progress" in paths) {
                builder.setProjectProgress(progress)
            }

            if ("creation_date" in paths) {
                builder.setCreationDate(timestamp { seconds = project.createdAt.toEpochSecond() })
            }

            if ("last_stage_started" in paths) {
                builder.setLastStageStarted(timestamp { seconds = project.currentStageStartedAt.toEpochSecond() })
            }

            builder.build()
```

Also, if no path is included in the field mask, we should return nothing, since we use the field mask as opt-in, or rather, we have done it like that until now.

Reply:
> What's your opinion on the following? The timestamp builders can be replaced with the DSL

Implemented this :+1: 

> and we can decrease the indentation by just storing and using the builder.

:+1:

> It's "last_stage_started" instead of "last_stage_created".

Thank you. :+1: 

> Furthermore, we can just check whether a path is in the paths set [...]
> Also, if no path is included in the field mask, we should return nothing, since we use the field mask as opt-in, or rather, > we have done it like that until now.

This, I don't agree with. IMO there is no sense in returning a response with absolutely no information other than that it succeeded. This is also not how we've implemented it in the frontend. I've adjusted the field-mask-validator to optionally allow empty field masks (off per default). This is non-breaking and allows every function to decide for itself. We've used field masks for update operations, and there the field mask should be required, but I do not see the use in enforcing this for a get operation.

Reply:
I declared my point of view in the conversation above. You can see my proposal in the PR of https://github.com/SE-UUlm/snowballr-api/issues/95. Close this if you've read it and are fine with it.